app_feup/lib/controller/local_storage/app_restaurant_database.dart

```diff
@@ -0,0 +1,117 @@
+import 'package:intl/intl.dart';
+import 'package:sqflite/sqflite.dart';
+import 'package:uni/model/entities/meal.dart';
+import 'package:uni/model/entities/restaurant.dart';
+import 'package:uni/model/utils/day_of_week.dart';
+
+import 'app_database.dart';
+
+class RestaurantDatabase extends AppDatabase {
+  RestaurantDatabase()
+      : super('restaurant.db',
+      ['CREATE TABLE RESTAURANTS(id INTEGER PRIMARY KEY, ref TEXT , name TEXT)',
+      '''CREATE TABLE MEALS(
+          id INTEGER PRIMARY KEY AUTOINCREMENT,
+          day TEXT,
+          type TEXT,
+          date TEXT,
+          name TEXT,
+          id_restaurant INTEGER,
+          FOREIGN KEY (id_restaurant) REFERENCES RESTAURANTS(id))''']);
+
+  /**
+   * Delets all data, and saves the new restaurants
+   */
+  void saveRestaurants(List<Restaurant> restaurants) async{
+    final Database db = await this.getDatabase();
+    //final Batch batch = db.batch();
+    db.transaction((transaction) async {
+      deleteAll(transaction);
+      restaurants.forEach((restaurant) {
+        insertRestaurant(transaction, restaurant);
+      });
+    });
+
+
+  }
+
+  /**
+   * Get all restaurants and meals, if day is null, all meals are returned
+   */
+  Future<List<Restaurant>> restaurants({DayOfWeek day = null}) async {
+    final Database db = await this.getDatabase();
+    final List<Map<String, dynamic>> restaurantMaps =
+      await db.query('restaurants');
+
+    final List<Restaurant> restaurants =
+      await Future.wait(restaurantMaps.map((map) async {
+        final int restaurantId =  map['id'];
+
+        final List<Meal> meals =
+          await getRestaurantMeals(db, restaurantId, day: day);
+
+        return Restaurant(restaurantId, map['name'], map['ref'], meals: meals);
+    }).toList());
+
+    return restaurants;
```

Comment:
Could this also make use of transactions?

Reply:
Like this? I don't know another way to make the transaction return the restaurants.

```suggestion
    final Database db = await this.getDatabase();
    List<Restaurant> restaurants;
    await db.transaction((txn)  async {
      final List<Map<String, dynamic>> restaurantMaps =
      await txn.query('restaurants');

      restaurants = await Future.wait(restaurantMaps.map((map) async {
        final int restaurantId =  map['id'];
        final List<Meal> meals =
          await getRestaurantMeals(txn, restaurantId, day: day);

        return Restaurant(restaurantId, map['name'], map['ref'], meals: meals);
      }).toList());

    });


    return restaurants;
```

Reply:
yeah, that looks great!